pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,327 @@
+ar.register_function("numpy", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("builtins", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("numpy", "gather", lambda x, indices: x[np.array(indices)])
+
+
+def _scatter_element_add_numpy(tensor, index, value):
+    tensor[tuple(index)] += value
+    return tensor
+
+
+ar.register_function("numpy", "scatter_element_add", _scatter_element_add_numpy)
+
+
+# -------------------------------- Autograd --------------------------------- #
+
+
+# When autoray inspects PennyLane NumPy tensors, they will be associated with
+# the 'pennylane' module, and not autograd. Set an alias so it understands this is
+# simply autograd.
+ar.autoray._BACKEND_ALIASES["pennylane"] = "autograd"
+
+# When dispatching to autograd, ensure that autoray will instead call
+# qml.numpy rather than autograd.numpy, to take into account our autograd modification.
+ar.autoray._MODULE_ALIASES["autograd"] = "pennylane.numpy"
+
+
+ar.register_function("autograd", "flatten", lambda x: x.flatten())
+ar.register_function("autograd", "coerce", lambda x: x)
+ar.register_function("autograd", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("autograd", "gather", lambda x, indices: x[np.array(indices)])
+
+
+def _to_numpy_autograd(x):
+    if hasattr(x, "_value"):
+        # Catches the edge case where the data is an Autograd arraybox,
+        # which only occurs during backpropagation.
+        return x._value
+
+    return x.numpy()
+
+
+ar.register_function("autograd", "to_numpy", _to_numpy_autograd)
+
+
+def _scatter_element_add_autograd(tensor, index, value):
+    size = tensor.size
+    flat_index = _i("qml").numpy.ravel_multi_index(index, tensor.shape)
+    t = [0] * size
+    t[flat_index] = value
+    return tensor + _i("qml").numpy.array(t).reshape(tensor.shape)
+
+
+ar.register_function("autograd", "scatter_element_add", _scatter_element_add_autograd)
+
+
+def _take_autograd(tensor, indices, axis=None):
+    indices = _i("qml").numpy.asarray(indices)
+
+    if axis is None:
+        return tensor.flatten()[indices]
+
+    fancy_indices = [slice(None)] * axis + [indices]
+    return tensor[tuple(fancy_indices)]
+
+
+ar.register_function("autograd", "take", _take_autograd)
+
+
+# -------------------------------- TensorFlow --------------------------------- #
+
+
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "angle"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arcsin"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arccos"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arctan"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "diag"] = "tensorflow.linalg"
+
+
+ar.autoray._FUNC_ALIASES["tensorflow", "arcsin"] = "asin"
+ar.autoray._FUNC_ALIASES["tensorflow", "arccos"] = "acos"
+ar.autoray._FUNC_ALIASES["tensorflow", "arctan"] = "atan"
+ar.autoray._FUNC_ALIASES["tensorflow", "diag"] = "diag"

```

Comment:
How come an alias is needed here? Would it be `tensorflow.diag` if it wasn't for creating this alias?

Reply:
Oh good question! I don't remember why this is here unfortunately